internal/abci/abci.go

```diff
@@ -476,11 +612,55 @@ func validatorUpdates(initial, final []*types.Validator) []*types.Validator {
 // Commit persists the state changes. This is called under mempool lock in
 // cometbft, unlike FinalizeBlock.
 func (a *AbciApp) Commit(ctx context.Context, _ *abciTypes.RequestCommit) (*abciTypes.ResponseCommit, error) {
-	height, err := a.txApp.Commit(ctx)
+	err := a.consensusTx.Commit(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to commit transaction app: %w", err)
+	}
+
+	// we need to re-open a new transaction just to write the apphash
+	// TODO: it would be great to have a way to commit the apphash without
+	// opening a new transaction. This could leave us in a state where data is
+	// committed but the apphash is not, which would essentially nuke the chain.
+	tx, err := a.db.BeginOuterTx(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin outer tx: %w", err)
+	}
+
+	err = meta.SetChainState(ctx, tx, a.height, a.appHash)
+	if err != nil {
+		err2 := tx.Rollback(ctx)
+		if err2 != nil {
+			return nil, fmt.Errorf("failed to rollback transaction app: %w", err2)
+		}
+		return nil, fmt.Errorf("failed to set chain state: %w", err)
+	}
+
+	err = tx.Commit(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to commit transaction app: %w", err)
 	}
-	a.height.Store(height)
+
+	a.txApp.Commit(ctx)
+
+	if a.snapshotter != nil && a.replayingBlocks != nil &&
+		a.snapshotter.IsSnapshotDue(uint64(a.height)) && !a.replayingBlocks() {
+		// we make a snapshot tx but don't directly use it. This is because under the hood,
+		// we are using the pg_dump executable to create the snapshot, and we are simply
+		// giving pg_dump the snapshot ID to guarantee it has an isolated view of the database.
+		snapshotTx, snapshotId, err := a.db.BeginSnapshotTx(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to begin snapshot tx: %w", err)
+		}
+		defer snapshotTx.Rollback(ctx) // always rollback, since this is just for view isolation
+
+		err = a.snapshotter.CreateSnapshot(ctx, uint64(a.height), snapshotId)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create snapshot: %w", err)
```

Comment:
Snapshot creation errors should be ignored, else the block production would stop. Just log the error and return nil to the cometbft

Reply:
In what cases would this be an error though?

Reply:
We actually just hit one when playing with the snapshots folder config setting:

on `main` it doesn't crash, only log:

```go
2024-06-17T22:08:22.186Z        error   kwild.tx-router txapp/txapp.go:816      failed to dump logical snapshot {"error": "failed to create snapshot at height 530: mkdir /data/kwil-snaps2/block-530: permission denied", "goversion": "go1.22.4", "mod_version": "(devel)", "revision": "6a09f17e03f532df2b3fa58ffb2c113b5d313779", "build_time": "2024-06-13T16:24:07Z", "dirty": tru
e}                                                                                                                                                                                          
github.com/kwilteam/kwil-db/internal/txapp.(*TxApp).snapshotDatabase                                                                                                                        
        /home/ubuntu/kwil/git/kwil-db/internal/txapp/txapp.go:816                                                                                                                           github.com/kwilteam/kwil-db/internal/txapp.(*TxApp).Commit                                                                                                                                  
        /home/ubuntu/kwil/git/kwil-db/internal/txapp/txapp.go:799                                                                                                                           github.com/kwilteam/kwil-db/internal/abci.(*AbciApp).Commit                                   
        /home/ubuntu/kwil/git/kwil-db/internal/abci/abci.go:479                                                                                                                             
github.com/cometbft/cometbft/abci/client.(*localClient).Commit                                                                                                                              
        /data/gopath/pkg/mod/github.com/cometbft/cometbft@v0.38.7/abci/client/local_client.go:113               
```

Reply:
```
snapshotTx, snapshotId, err := a.db.BeginSnapshotTx(ctx)
		if err != nil {
			return nil, fmt.Errorf("failed to begin snapshot tx: %w", err)
		}
```

Lets remove this too. Essentially intent is the node should produce blocks even if snapshot creation fails.

Reply:
I see. Ok.